### ISSUE TITLE
How to set a custom input timestamp in evaluation

### ISSUE THREAD

**Issue by szl0144** [nianticlabs/panoptic-forecasting#4]
Hi,

I see the current short-term and long-term evaluation input timestamp is fixed (Short term: 11, 14, 17, Long term: 5 8 11, label frame is 20th). Can you please tell me which code I need to focus on to set a custom input timestamp besides short term and long term? 

I see the model forecasts future 3 frames in short term evaluation and 9 frames in long term evaluation, does the loss function only calculate the SQ, RQ and PQ between 3rd prediction frame and the 20th ground truth frame in short term and the loss between 9th prediction frame and the 20th ground truth in long term? The loss function doesn't calculate the loss of other prediction frames (such as 4th, 5th, 6th in this scenario) because of a lack of ground truth. Not sure about it and want to double-check it with you. Thanks!

Best,
Shaoyi

**Comment by cgraber** [comment]
Your understanding of the code is correct - since there isn't ground-truth for any frame besides the 20th in Cityscapes, all evaluations are done on this frame. This code is set up with the fixed assumption that you want one of the two settings (short or mid-term) that were used for the paper. It is possible to set it up using other settings, but it will take some work.

The easiest part of the model to train on a different length of time is the foreground model. [These lines of code](https://github.com/nianticlabs/panoptic-forecasting/blob/master/panoptic_forecasting/data/datasets/fg_instance_dataset.py#L155-L173) in the instance dataloader specify the time intervals which are used for training; you should be able to work from there to figure out how to change the lengths of time considered for both the input and forecast.

The background component takes a bit more work by nature of the data preprocessing done. The code assumes the projection of the input class pixels has already been done, and it learns to predict a full background semantic segmentation prediction using these reprojected point clouds as input. The code to do the reprojection [is in the repo](https://github.com/nianticlabs/panoptic-forecasting/blob/master/panoptic_forecasting/models/pc_transform/pc_transform_model.py), but I haven't provided information regarding how to actually run this. That's because the reprojection step requires running several models on input data to get intermediate predictions that are then used, including semantic segmentations for input frames as well as depth prediction.

Fully documenting this is something that I haven't done yet, but I will start working on it now for the benefit of people who may be interested in either using different depths/semantic segmentation inputs or who are interested in using different input/output sequence lengths

**Comment by szl0144** [comment]
Hi,

Make sense, thanks for your detailed explanation! it seems I can start from the foreground model first and wait for the introduction to run the models of cloud point reprojections. It will be great if you can provide more documents I can follow, thanks!

One more question, if I want to predict future frames with custom lengths, need I do any revision in the prediction script or code? Or just need to change the codes in the training part and keep the prediction part.

**Comment by szl0144** [comment]
Hi,

Sorry, I just reviewed the suggestions you give me and I think I didn't describe my question clearly. I would like to predict a custom-term frame such as future 8 frames (the input sequence should be 6,9,12) or future 2 frames (the input should be 12,15,18) with the same length interval when the model is trained which is 3 frames. Can you please tell me which part I need to focus on for the custom-term frame prediction?

**Comment by cgraber** [comment]
I'm not sure I 100% understand the question, so I'll do my best here and you can clarify if I misunderstood:

First, answering your previous question: if you want to change either the input or target sequence for prediction, you'll need to make changes for both inference and training. The place I pointed to you earlier should be all you need to change for training purposes, as the model will adapt to whatever input/output sequence lengths you specify there.

If the goal is to change the frames used for prediction during inference, look [here](https://github.com/nianticlabs/panoptic-forecasting/blob/master/panoptic_forecasting/data/datasets/fg_scene_dataset.py#L185-L191). Specifically, the `base_inds` array specifies the spacing of frames provided as input, while `start_inds` specifies the index of the last frame provided as input.

To change the amount of "steps" that are predicted during inference, look [here](https://github.com/nianticlabs/panoptic-forecasting/blob/master/panoptic_forecasting/models/fg/fg_model.py#L529). The amount of steps you want to predict is passed as input to the forward function; this is currently set to the length of the annotation sequence, but you could set it to some other length if you'd like.

**Comment by szl0144** [comment]
Hii,
Thanks for your explanations, got it. I think the code locations you mark are what I need.
Can you double check if my understanding is correct?
What I want to do is to predict the loss of future 10 images. Because only the 20th frame has a label, I need to adapt the input sequence frame index. 
For example, I need to use frames (4,7,10) to predict the future 10 frames and calculate the loss of the 10th frame.
I need to use frames (5,8,11) to predict the future 9 frames and calculate the loss of the 9th frame.
I need to use frames (6,9,12) to predict the future 8 frames and calculate the loss of the 8th frame.
I need to use frames (7,10,13) to predict the future 7 frames and calculate the loss of the 7th frame.
I need to use frames (8,11,14) to predict the future 6 frames and calculate the loss of the 6th frame.
I need to use frames (9,12,15) to predict the future 5 frames and calculate the loss of the 5th frame.
I need to use frames (10,13,16) to predict the future 4 frames and calculate the loss of the 4th frame.
I need to use frames (11,14,17) to predict the future 3 frames and calculate the loss of the 3rd frame.
I need to use frames (12,15,18) to predict the future 2 frames and calculate the loss of the 2nd frame.
I need to use frames (13,16,19) to predict the future 1 frames and calculate the loss of the 1st frame.

This is the plan I want to do.

**Comment by szl0144** [comment]
According to your suggestions, I don't need to change the ```base_inds``` because the spacing is still 3. 

I need to change the ```start_inds``` to get the last index in the input. For example, If I want to predict the future 9th frame, I need to change the last index to 11 in the input (5,8,11).
I also need to change the future steps to the frame index that I need. For example, , If I want to predict the future 9th frame, I need to change the steps to 9.

Please correct me if I misunderstand it. Thanks!

**Comment by cgraber** [comment]
I think your understanding is close. Note that the base inds contains the indices for both the past (i.e. input) frames as well as the future frames to be predicted. So, as an example, if you want to change the frames used during training to have different spacing, you'll need to adjust the base inds as well. The dataloaders load the information for the specified frames and then splits them into "input" and "target" based on the specified input len (which, e.g., happens [here](https://github.com/nianticlabs/panoptic-forecasting/blob/master/panoptic_forecasting/data/datasets/fg_instance_dataset.py#L343) for the training version of the dataloader)

Currently, the model is trained such that it predicts every 3rd frame. Hence, setting steps to 3 predicts the 3rd, 6th, and 9th frames into the future. If you want a model that can predict at different intervals (or more than 3 steps == 9 frames into the future), you'll need to train a new model, since the provided pretrained model was not trained on this setting.

**Comment by szl0144** [comment]
Hi,
Thanks for your help. I think I got your point and know how to make a prediction with custom prediction length. I also print out the variables you mentioned to have a better understanding. These are values I print.
1) ```base_inds``` = [ 0  3  6  9 12 15]
2) ```start_inds``` = 4
3) ```out_t``` = label_trajs.size(1)= 3

As you said, the '''base_inds''' contains both input and output frames, and the current model predicts the 3rd, 6th, and 9th frames. Do you mean the frames indices for input is [12,15,0] and output is [3,6,9]? So the '''start_inds''' = 4 is base_inds[4] =12 which is the first element of input. Sorry I am confused about this part. 

One more question is I notice there are two py files about the sequence input and otput, they are fg_instance_dataset.py and fg_scene_dataset. Need I revise both of the two files for the prediction part?

Best,
Shaoyi

**Comment by cgraber** [comment]
The frames that are presented to the model are base_inds + start_ind. So, for the example you listed, the final frames that are used are [0 3 5 9 12 15] + 4 = [4, 7, 10, 13, 16, 19].

Regarding your second question: fg_instance_dataset.py is used during training, and samples individual trajectories to present to the model. fg_scene_dataset.py is used during evaluation, and presents all trajectories for a given scene at once. So if you're training, you'll want to change both, but if you're only evaluating, you only need to change fg_scene_dataset.

**Comment by szl0144** [comment]
Hi,

Make sense, so the [4,7,10] is the input sequence and the output sequence is [13,16,19]. The script I run is the mid term script, and I see the input sequence of mid term in the paper is [5,8,11]. Just want to double check the code in Github is different from the code used for the paper.

One more question I am confused is about the loss calculation. We get the output sequence [13,16,19] but none of them is at 20 which is the index of ground truth. Can you please tell me do you calculate the loss with the output sequence [13,16,19]? Thanks!

Best,
Shaoyi

**Comment by cgraber** [comment]
This is just a matter of 0-indexing vs. 1-indexing. The description in the paper is 1-indexed (i.e. the first frame in an entire sequence is frame 1 and the target frame is frame 20), while the code is 0-indexed (i.e. the first frame in an entire sequence is frame 0 and the target frame is frame 19). This explains the difference between what's in the code and what's written in the paper.

**Comment by szl0144** [comment]
Got it, thanks! Can you tell me how I can check the indexing type of the code? It is related to dataset loader settings and it won't be changed if I didn't make change in that settings part, right?

**Comment by cgraber** [comment]
This isn't changeable in the code at all, actually - everything is written with the assumption that sequences are zero-indexed. This matches how cityscapes files are indexed in practice as well - for example, sequences from lindau all start from frame 0 and the 20th frame of the sequence is therefore frame 19.

**Comment by szl0144** [comment]
Make sense, I misunderstand the indexing type of the dataset. I thought it's from 1 rather than 0, thanks for your explanations. Based on your help, I think I need to change the dataset loader in both the training and evaluation code, and train the model again to make it predict future frames at other timestamps besides the timestamp in short term and long term evaluation.

One question I have is whether I must keep one frame index 19 in the array ```current_inds```= ```base_inds```+```start_inds``` no matter where the frame index 19 is in the output sequence.  For example, I can use [0 3 6 9 12 15] + 7=[7,10,13,16,19,22] for loss calculation where the index19 is the second frame in the output sequence. I also notice the index in the output sequence can be over 19  (the index 22 in the example), can the index in the input sequence smaller than 0? For example, can I use ```base_inds``` = np.arange(-1, 3*(self.input_len + self.output_len), 3)?

**Comment by cgraber** [comment]
This won't work on the default cityscapes dataset, as the code will attempt to load a file that doesn't exist (for all indices less than 0). There are 30 frames per sequence for cityscapes, which is why frames larger than 19 work fine (they can go up to 29).
